pkg/river/internal/value/decode.go

```diff
@@ -0,0 +1,382 @@
+	//
+	// NOTE(rfratto): we don't reassign to val here, since Go 1.18 thinks that
+	// means it escapes the heap. We need to create a local variable to avoid
+	// exctra allocations.
+	convVal := val
+
+	// Convert the value.
+	switch {
+	case val.rv.Type() == goByteSlice && into.Type() == goString: // []byte -> string
+		into.Set(val.rv.Convert(goString))
+		return nil
+	case val.rv.Type() == goString && into.Type() == goByteSlice: // string -> []byte
+		into.Set(val.rv.Convert(goByteSlice))
+		return nil
+	case convVal.Type() != targetType:
+		var err error
+		convVal, err = convertValue(convVal, targetType)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Slowest case: recursive decoding. Once we've reached this point, we know
+	// that convVal.rv and into are compatible Go types.
+	switch convVal.Type() {
+	case TypeNumber:
+		into.Set(convertGoNumber(convVal.rv, into.Type()))
+		return nil
+	case TypeString:
+		into.Set(convVal.rv)
+		return nil
+	case TypeBool:
+		into.Set(convVal.rv)
+		return nil
+	case TypeArray:
+		return decodeArray(convVal, into)
+	case TypeObject:
+		return decodeObject(convVal, into)
+	case TypeFunction:
+		// If the function types had the exact same signature, they would've been
+		// handled in the best case statement above. If we've hit this point,
+		// they're not the same.
+		//
+		// For now, we return an error.
+		//
+		// TODO(rfratto): we may want to consider being more lax here, potentially
+		// creating an adapter between the two functions.
+		return DecodeError{
+			Value: val,
+			Inner: fmt.Errorf("expected function(%s), got function(%s)", into.Type(), convVal.rv.Type()),
+		}
+	case TypeCapsule:
+		// Capsule types require being identical go types, which would've been
+		// handled in the best case statement above. If we hit this point, they're
+		// not the same.
+		//
+		// TODO(rfratto): return a TypeError for this instead. TypeError isn't
+		// appropriate at the moment because it would just print "capsule", which
+		// doesn't contain all the information the user would want to know (e.g., a
+		// capsule of what inner type?).
+		return DecodeError{
+			Value: val,
+			Inner: fmt.Errorf("expected capsule(%s), got %s", into.Type(), convVal.Describe()),
+		}
+	default:
+		panic("river/value: unexpected kind " + convVal.Type().String())
+	}
+}
+
+func decodeArray(val Value, rt reflect.Value) error {
+	switch rt.Kind() {
+	case reflect.Slice:
+		res := reflect.MakeSlice(rt.Type(), val.rv.Len(), val.rv.Len())
+		for i := 0; i < val.rv.Len(); i++ {
+			// Decode the original elements into the new elements.
+			if err := decode(val.Index(i), res.Index(i)); err != nil {
+				return ElementError{Value: val, Index: i, Inner: err}
+			}
+		}
+		rt.Set(res)
+
+	case reflect.Array:
+		res := reflect.New(rt.Type()).Elem()
+		for i := 0; i < val.rv.Len(); i++ {
+			// Stop processing elements if the target array is too short.
+			// TODO(rfratto): should we force array length to be identical?
```

Comment:
Better than silently dropping elements IMHO. I think if someone's going through the trouble of using a Go array, they'd have a reason for wanting exactly N elements.

Reply:
Yeah, I agree. Silently dropping is the behavior encoding/json uses, but I find it weird. I'll change it to enforce the length.